src/goto-programs/goto_convert_side_effect.cpp

```diff
@@ -99,33 +102,19 @@ void goto_convertt::remove_assignment(
 
     const typet &op0_type = expr.op0().type();
 
-    if(op0_type.id()==ID_c_bool)
-    {
-      // C/C++ Booleans get very special treatment.
-      binary_exprt tmp(expr.op0(), new_id, expr.op1(), expr.op1().type());
-      tmp.op0().make_typecast(expr.op1().type());
-      rhs=typecast_exprt(is_not_zero(tmp, ns), expr.op0().type());
-    }
-    else if(op0_type.id() == ID_c_enum_tag)
-    {
-      // We convert c_enums to their underlying type, do the
-      // operation, and then convert back
-      const auto &enum_type = ns.follow_tag(to_c_enum_tag_type(op0_type));
-      auto underlying_type = enum_type.subtype();
-      auto op0 = typecast_exprt(expr.op0(), underlying_type);
-      auto op1 = typecast_exprt(expr.op1(), underlying_type);
-      binary_exprt tmp(op0, new_id, op1, underlying_type);
-      rhs = typecast_exprt(tmp, expr.op0().type());
-    }
-    else
-    {
-      rhs.id(new_id);
-      rhs.copy_to_operands(expr.op0(), expr.op1());
-      rhs.type()=expr.op0().type();
-      rhs.add_source_location()=expr.source_location();
-    }
+    PRECONDITION(
+      op0_type.id() != ID_c_enum_tag && op0_type.id() != ID_c_enum &&
+      op0_type.id() != ID_c_bool && op0_type.id() != ID_bool);
+    rhs.id(new_id);
+    rhs.copy_to_operands(expr.op0(), expr.op1());
+    rhs.type() = expr.op0().type();
+    rhs.add_source_location() = expr.source_location();
+
+    exprt new_lhs = skip_typecast(expr.op0());
+    rhs = typecast_exprt::conditional_cast(rhs, new_lhs.type());
+    rhs.add_source_location() = expr.source_location();
```

Comment:
Same here.

Reply:
This is outside the C front-end, I don't think I can use `implicit_typecast(_arithmetic)` here?!

Reply:
Yes, you may have to add `ID_C_implicit_typecast` directly.